From: Julian Ospald <hasufell@gentoo.org>
Date: Mon Aug 20 17:16:06 UTC 2012
Subject: build system

respect users flags

--- kmscon/Makefile.am
+++ kmscon/Makefile.am
@@ -61,8 +61,6 @@
 	-include $(top_builddir)/config.h \
 	-I $(srcdir)/src \
 	-I $(srcdir)/external
-AM_LDFLAGS = \
-	-Wl,--as-needed
 
 if DEBUG
 AM_CFLAGS += -g
@@ -70,8 +68,6 @@
 
 if OPTIMIZATIONS
 AM_CFLAGS += -O2
-else
-AM_CFLAGS += -O0
 endif
 
 #
