--- vfe/vfe.cpp    2009-02-26 02:09:03.000000000 +0300
+++ vfe/vfe.cpp    2010-01-13 15:07:40.000000000 +0300
@@ -748,8 +748,8 @@
   // this code may not be removed or modified unless the compilation is for personal
   // use and the binary is not distributed to anyone else.
   xtime_get(&t, TIME_UTC);
-  if (t.sec > 1262296800) // 1 Jan 2010
-    throw POV_EXCEPTION(kCannotHandleRequestErr, "Please obtain an updated version of this application.");
+  // if (t.sec > 1262296800) // 1 Jan 2010
+  //  throw POV_EXCEPTION(kCannotHandleRequestErr, "Please obtain an updated version of this application.");

   state = kStarting;


