diff -urN gosa-core.orig/include/class_config.inc gosa-core/include/class_config.inc
--- gosa-core.orig/include/class_config.inc	2011-07-27 10:38:29.000000000 +0400
+++ gosa-core/include/class_config.inc	2011-11-10 00:05:10.583476711 +0400
@@ -116,12 +116,12 @@
            if we have installed or removed plugins. 
          */
         if(session::global_is_set("class_location.inc:timestamp")){
-            $tmp = stat("../include/class_location.inc");
+	    $tmp = $tmp = stat("/var/cache/gosa/class.cache");
             if($tmp['mtime'] != session::global_get("class_location.inc:timestamp")){
                 session::global_un_set("plist");
             }
         }
-        $tmp = stat("../include/class_location.inc");
+        $tmp = stat("/var/cache/gosa/class.cache");
         session::global_set("class_location.inc:timestamp",$tmp['mtime']);
 
         if($this->filename != "" && filemtime($this->filename) != $this->last_modified){
diff -urN gosa-core.orig/include/functions.inc gosa-core/include/functions.inc
--- gosa-core.orig/include/functions.inc	2011-10-05 17:35:02.000000000 +0400
+++ gosa-core/include/functions.inc	2011-11-10 00:06:18.908744189 +0400
@@ -44,7 +44,7 @@
 }
 
 /* Define common locatitions */
-define ("CONFIG_TEMPLATE_DIR", "../contrib");
+define ("CONFIG_TEMPLATE_DIR", "/usr/share/doc/gosa");
 define ("TEMP_DIR","/var/cache/gosa/tmp");
 
 /* Define get_list flags */
@@ -73,7 +73,7 @@
 define('DES3_CBC_SHA1',16);
 
 /* Include required files */
-include_once("class_location.inc");
+include_once("/var/cache/gosa/class.cache");
 require_once ("functions_debug.inc");
 require_once ("accept-to-gettext.inc");
 
diff -urN gosa-core.orig/include/php_setup.inc gosa-core/include/php_setup.inc
--- gosa-core.orig/include/php_setup.inc	2011-10-25 17:00:11.000000000 +0400
+++ gosa-core/include/php_setup.inc	2011-11-10 00:06:46.598853022 +0400
@@ -20,7 +20,7 @@
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
  */
 
-define ("LOCALE_DIR", dirname(dirname(__FILE__))."/locale/compiled");
+define ("LOCALE_DIR", "/var/cache/gosa/locale");
 
 function gosaRaiseError($errno, $errstr, $errfile, $errline)
 {
diff -urN gosa-core.orig/update-gosa gosa-core/update-gosa
--- gosa-core.orig/update-gosa	2010-12-06 10:52:44.000000000 +0300
+++ gosa-core/update-gosa	2011-11-10 00:06:46.598853022 +0400
@@ -21,7 +21,7 @@
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
  */
 
-define ("GOSA_HOME", dirname(__FILE__));
+define ("GOSA_HOME", "/usr/share/gosa");
 define ("LOCALE_DIR", GOSA_HOME."/locale");
 define ("PLUGSTATE_DIR", GOSA_HOME."/state");
 
@@ -138,8 +138,8 @@
 function rescan_classes()
 {
 	echo "Updating class cache...\n";
-	$class_mapping= get_classes();
-	$filename= GOSA_HOME."/include/class_location.inc";
+	$class_mapping= get_classes("/usr/share/gosa");
+	$filename= "/var/cache/gosa/class.cache";
 
 	/* Sanity checks */
 	if (!file_exists($filename) || is_writable($filename)) {
@@ -185,20 +185,20 @@
 
 	/* For each language, merge the target .mo to the compiled directory. */
 	foreach ($languages as $language => $po_files){
-		if (!is_dir(LOCALE_DIR."/compiled/${language}/LC_MESSAGES")){
-			if (!mkdir (LOCALE_DIR."/compiled/${language}/LC_MESSAGES", 0755, TRUE)){
-				echo "Failed to create '".LOCALE_DIR."/compiled/${language}/LC_MESSAGES'- aborted";
+		if (!is_dir("/var/cache/gosa/locale/${language}/LC_MESSAGES")){
+			if (!mkdir ("/var/cache/gosa/locale/${language}/LC_MESSAGES", 0755, TRUE)){
+				echo "Failed to create '/var/cache/gosa/locale/${language}/LC_MESSAGES'- aborted";
 				exit (3);
 			}
 		}
 
 		/* Cat all these po files into one single file */
-		system ("(cd ".LOCALE_DIR." && msgcat --use-first ".implode(" ", $po_files)." > compiled/${language}/LC_MESSAGES/messages.po)", $val);
+		system ("(cd ".LOCALE_DIR." && msgcat --use-first ".implode(" ", $po_files)." > /var/cache/gosa/locale/${language}/LC_MESSAGES/messages.po)", $val);
 		if ($val != 0){
 			echo "Merging of message files failed - aborted";
 			exit (4);
 		}
-		system ("(cd ".LOCALE_DIR."/compiled/${language}/LC_MESSAGES && msgfmt -o messages.mo messages.po && rm messages.po)", $val);
+		system ("(cd /var/cache/gosa/locale/${language}/LC_MESSAGES && msgfmt -o messages.mo messages.po && rm messages.po)", $val);
 		if ($val != 0){
 			echo "Compiling of message files failed - aborted";
 			exit (5);
