diff --git a/src/lxc/confile.c b/src/lxc/confile.c
index c217e19..1e50952 100644
--- a/src/lxc/confile.c
+++ b/src/lxc/confile.c
@@ -39,6 +39,7 @@
 
 lxc_log_define(lxc_confile, lxc);
 
+static int config_include(const char *, char *, struct lxc_conf *);
 static int config_pts(const char *, char *, struct lxc_conf *);
 static int config_tty(const char *, char *, struct lxc_conf *);
 static int config_cgroup(const char *, char *, struct lxc_conf *);
@@ -63,7 +64,7 @@ struct config {
 };
 
 static struct config config[] = {
-
+	{ "config.include",     config_include        },
 	{ "lxc.pts",            config_pts            },
 	{ "lxc.tty",            config_tty            },
 	{ "lxc.cgroup",         config_cgroup         },
@@ -529,6 +530,12 @@ static int config_utsname(const char *key, char *value, struct lxc_conf *lxc_con
 	return 0;
 }
 
+
+static int config_include(const char *key, char *value, struct lxc_conf *lxc_conf)
+{
+	return lxc_config_read( value, lxc_conf );
+}
+
 static int parse_line(void *buffer, void *data)
 {
 	struct config *config;
